### ISSUE TITLE
Solution design using ImageSharp in Greenshot

### ISSUE THREAD

**Issue by Lakritzator** [greenshot/greenshot#364]
**Is your feature request related to a problem? Please describe.**
With PR #358 we might introduce ImageSharp, and derivates (ImageSharp.Drawing).
It makes sense to describe the way to go forward with this, make sure we have guidelines.

**Describe the solution you'd like**
That is what is going to be in the comments following this.

**Describe alternatives you've considered**
For the PR itself, Emoji.Wpf, but that would be a very specific solution for one single feature.

**Additional context**
I see a future using ImageSharp

**Comment by Lakritzator** [comment]
This issue was created due to the PR mentioned in the initial comment, for this the requirements are:

1. Draw an emoji to a WPF bitmap (System.Windows.Media.Imaging.BitmapSource) so it can be displayed in a picker, which is currently implemented in WPF.
2. Draw an emoji to the Greenshot canvas, which is a System.Drawing.Bitmap, so the user can draw with it.

SixLabors.ImageSharp.Drawing can draw text (with specific fonts which in the case of the PR), which is used to draw the emoji.
As we already have a deeply nested System.Drawing usage, we cannot just witch to ImageSharp.

To cover the requirements we need at least a way to convert or wrap the BitmapSource or Bitmap in such a way we can call DrawText.

**Comment by Lakritzator** [comment]
Another requirement that is currently seen as a low hanging fruit:
- Add additional, or better, support for file formats (e.g. webp which is requested). Both for saving and loading.

**Comment by Lakritzator** [comment]
Furthermore I can image:

- Replacing the resize and quantizers with the very good quality of ImageSharp.
- Replace GDI DrawText, so we are no longer hitting restrictions like #283
- Move the effects to ImageSharp based logic, making platform agnostic usage possible

I currently cannot image that ImageSharp replaces:

- The bitmap animations on the capture form or in the about box.

Doubtful, but who knows:
- Migrate the surface (the drawing in the editor) drawing code to ImageSharp.

**Comment by Lakritzator** [comment]
Restrictions / facts:

- ImageSharp has a mutate, and allows you to resize the image in place. For Greenshot there are two modes of operandum, one is drawing to a bitmap, the other one is creating a new bitmap (used when resizing). There is no easy way to restrict the usage of ImageSharp, which could impact the effect in such a way it will break anything.
- Before using ImageSharp, one needs to consider the way it does memory management: https://docs.sixlabors.com/articles/imagesharp/memorymanagement.html

Suggestion:

- provide extension methods similar to the ones already available on Image<> to Bitmap and WritableBitmapSource. Create a Mutate, which checks if the image is not modified in such a way it's structure is modified (resize, other format). And create a Clone which allows this.

Open questions:

- What version(s) should we use?
- Does the license model fit with that of Greenshot?

**Comment by jairbubbles** [comment]
For file formats one can also hope that it provides a better quality / compression. (For instance .png produced by Greenshot are really huge)

**Comment by jairbubbles** [comment]
For alternatives : [Skia](https://skia.org/) and [Harfbuzz](https://github.com/harfbuzz/harfbuzz) for the font layouting but both are native libraries.

**Comment by Lakritzator** [comment]
> For file formats one can also hope that it provides a better quality / compression. (For instance .png produced by Greenshot are really huge)

That would be a nice and easy test to make, writing the same Bitmap as .PNG with System.Drawing and ImageSharp.
Maybe I should add the needed "core" code to one of my other projects, Dapplo.Windows, maybe call it Dapplo.Windows.ImageSharpInterop.

**Comment by jairbubbles** [comment]
I quickly tested saving in .png with ImageSharp and on my one example the file was bigger. This also made me realize that we do have a stride issue when dealing with 24 bits bitmap, the `WrapMemory` does expect the pixels to be contiguous whilst with bitmaps the stride must be a multiple of 4. (But with 32 bits images we don't have the issue then.)

**Comment by Lakritzator** [comment]
I know, I really expect a stride in the API, otherwise it won't work.

**Comment by Lakritzator** [comment]
So I am building some experimental code for the interop.
For the use-case image formats I just tested saving with an example photo: https://www.dpreview.com/sample-galleries/5902531902/leica-m11-sample-gallery/8521206921

The file sizes for this are:
With GDI (bitmap.Save): 2.319.709 bytes
PNG and ImageSharp SaveAsPng; 1.853.471 bytes

That is quite a huge difference!

https://github.com/dapplo/Dapplo.Windows/blob/feature/ImageSharpExperiment/src/Dapplo.Windows.Tests/ImageSharpInteropTests.cs#L67

**Comment by Lakritzator** [comment]
> Would be nice to see how much https://pngquant.org/ can gain on top of that!
That uses quantizing, reducing the colors to 256, is not directly a fair comparison.
The resulting file is ~772Kb

But I could probably modify the code with ImageSharp, quantize the image to 8 bit and save, it should be in the same area of bytes.

> ImageSharp seems to support less format than GDI.
I do not know what you mean, less formats? It has pretty much the same, I think, and it adds webp.

**Comment by jairbubbles** [comment]
> But I could probably modify the code with ImageSharp, quantize the image to 8 bit and save, it should be in the same area of bytes.

In my case I often make screenshots of UI which have not many colors so it's huge gain.

> I do not know what you mean, less formats? It has pretty much the same, I think, and it adds webp.

I clearly spoke too fast 😳 It's nice for .webp, it's not yet in the doc as it's quite recent. With WPF you can have .wepb but you need to install an extension in Microsoft Store.

**Comment by Lakritzator** [comment]
I'm thinking about building a **small** repository to evaluate different solutions, concerning image formats.

- System.Drawing (Windows)
- WPF (Windows)
- ImageSharp (Multi platform - no UI framework)

Optionally:
- Skia (Multi platform - Avalonia)
- WIC (Windows - no UI framework)

My goal is to structurally collect a bit of information for me, but also to support others with their decision.
The goal is not to get a complete overview over all features and solutions, but it's open for PR's 🤷

**Comment by Lakritzator** [comment]
True, WPF is a "thin" layer upon WIC. But it's also a whole UI framework. You could do WIC without the UI...
If you are not planning on using the UI, you get a lot of dependencies. I'm also not sure how WPF works with not pre defined encoders, but it somehow should work with .webp or .heic etc, if the codes are installed.

**Comment by jairbubbles** [comment]
> If you are not planning on using the UI, you get a lot of dependencies

Yes but WPF is included in .NET if you target Windows so there's not much gain here.

> but it somehow should work with .webp or .heic etc, if the codes are installed.

I succesfully used .webp but you need users to install it from the Microsoft Store so it's not handy at all.

**Comment by dannysummerlin** [comment]
Hi all, just to nudge this, is there a plan for bringing in ImageSharp or some alternative? It looks like the emoji work has stalled, I was thinking of starting to add a WebP handler with ImageSharp but then found this thread and saw it is a bigger question. Any news? Thanks!
